src/nf/nf_optimizers.f90

```diff
@@ -87,6 +87,20 @@ end subroutine minimize
     procedure :: minimize => minimize_adam
   end type adam
 
+  type, extends(optimizer_base_type) :: adagrad
+    !! Adagrad optimizer by Duchi et al. (2011)
+    !!
+    !! Duchi, J., Hazan, E. and Singer, Y., 2011. Adaptive subgradient
+    !! methods for online learning and stochastic optimization. Journal
+    !! of Machine Learning Research, 12(Jul), pp.2121-2159.
+    !! http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf
+    real :: epsilon = 1e-8
+    real, allocatable, private :: sum_squared_gradient(:)
```

Comment:
Need `learning_rate_decay` here ($\eta$ in https://pytorch.org/docs/stable/generated/torch.optim.Adagrad.html#torch.optim.Adagrad)

Reply:
Both Keras and PyTorch implement the `weight_decay` (the L2 kind) in Adagrad. We may as well do that since we're already here.